substrate/frame/revive/src/evm/runtime.rs

```diff
@@ -0,0 +1,669 @@
+	type SignedExtensions = E::Extra;
+}
+
+impl<Call: TypeInfo, E: EthExtra> ExtrinsicCall for UncheckedExtrinsic<Call, E> {
+	fn call(&self) -> &Self::Call {
+		self.0.call()
+	}
+}
+
+impl<Call, E, Lookup> Checkable<Lookup> for UncheckedExtrinsic<Call, E>
+where
+	Call: Encode + Member,
+	E: EthExtra,
+	<E::Config as frame_system::Config>::Nonce: TryFrom<U256>,
+	BalanceOf<E::Config>: Into<U256> + TryFrom<U256>,
+	MomentOf<E::Config>: Into<U256>,
+
+	AccountIdOf<E::Config>: Into<AccountId32>,
+	Call: From<crate::Call<E::Config>> + TryInto<crate::Call<E::Config>>,
+	E::Extra: SignedExtension<AccountId = AccountId32>,
+	Lookup: traits::Lookup<Source = MultiAddress, Target = AccountId32>,
+{
+	type Checked = CheckedExtrinsic<AccountId32, Call, E::Extra>;
+
+	fn check(self, lookup: &Lookup) -> Result<Self::Checked, TransactionValidityError> {
+		if self.0.signature.is_none() {
+			if let Ok(call) = self.0.function.clone().try_into() {
+				if let crate::Call::eth_transact {
+					payload,
+					gas_limit,
+					storage_deposit_limit,
+					transact_kind,
+				} = call
+				{
+					let checked = E::try_into_checked_extrinsic(
+						payload,
+						gas_limit,
+						storage_deposit_limit,
+						transact_kind,
+					)?;
+					return Ok(checked)
+				};
+			}
+		}
+		self.0.check(lookup)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn unchecked_into_checked_i_know_what_i_am_doing(
+		self,
+		lookup: &Lookup,
+	) -> Result<Self::Checked, TransactionValidityError> {
+		self.0.unchecked_into_checked_i_know_what_i_am_doing(lookup)
+	}
+}
+
+impl<Call, E> GetDispatchInfo for UncheckedExtrinsic<Call, E>
+where
+	Call: GetDispatchInfo,
+	E: EthExtra,
+{
+	fn get_dispatch_info(&self) -> DispatchInfo {
+		self.0.get_dispatch_info()
+	}
+}
+
+impl<Call: Encode, E: EthExtra> serde::Serialize for UncheckedExtrinsic<Call, E> {
+	fn serialize<S>(&self, seq: S) -> Result<S::Ok, S::Error>
+	where
+		S: ::serde::Serializer,
+	{
+		self.0.serialize(seq)
+	}
+}
+
+impl<'a, Call: Decode, E: EthExtra> serde::Deserialize<'a> for UncheckedExtrinsic<Call, E> {
+	fn deserialize<D>(de: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'a>,
+	{
+		let r = sp_core::bytes::deserialize(de)?;
+		Decode::decode(&mut &r[..])
+			.map_err(|e| serde::de::Error::custom(sp_runtime::format!("Decode error: {}", e)))
+	}
+}
+
+/// A [`SignedExtension`] that performs pre-dispatch checks on the Ethereum transaction's fees.
+#[derive(DebugNoBound, DefaultNoBound, Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct CheckEthTransact<T: Config> {
+	/// The gas fee, computed from the signed Ethereum transaction.
+	///
+	/// # Note
+	///
+	/// This is marked as `#[codec(skip)]` as this extracted from the Ethereum transaction and
+	/// should have a zero size encoded representation.
+	#[codec(skip)]
+	eth_fee: Option<BalanceOf<T>>,
+	_phantom: core::marker::PhantomData<T>,
+}
```

Comment:
It causes quite the code churn to add this fake signed extension everwhere. Why not to the `eth_fee` check directly in `Checkable::check`?

Reply:
I did not have that initially, but then realized I needed the encoded_len of the extrinsic to properly estimate the fees. We could use [encoded_size](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/trait.Encode.html#method.encoded_size) to recalculate it, but not sure this is much better 


>  Its also super confusing to add a SignedExtension specifically for an unsigned transaction

SignedExtension can be used for all transaction (all methods on the trait have _unsigned variant). 

Relevant code running this logic:
https://github.com/paritytech/polkadot-sdk/blob/pg/revive-rpc-server/substrate/frame/executive/src/lib.rs?plain=1#L329-L349

Reply:
> I did not have that initially, but then realized I needed the encoded_len of the extrinsic to properly estimate the fees. We could use [encoded_size](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/trait.Encode.html#method.encoded_size) to recalculate it, but not sure this is much better

I think it would make the code much easier to understand in my mind.

Reply:
I think it would make the code easier to understand. Not having this additional signed extension.

Reply:
Ok I will need to move a bunch of constraint but I'll give it a shot,
It just seems a bit wasteful to get get the dispatch_info and recompute the length here, when all of this is provided to the signed_extension later on.

Reply:
done the changes here https://github.com/paritytech/polkadot-sdk/pull/5866/commits/222a557ef22dceb9c3d8499c20cd0406e6d61645